core/src/main/scala/org/broadinstitute/dsde/rawls/fastpass/FastPassService.scala

```diff
@@ -205,175 +204,163 @@ class FastPassService(protected val ctx: RawlsRequestContext,
+      case SamWorkspaceRoles.reader | SamWorkspaceRoles.shareReader => Set(terraBucketReaderRole)
+      case _                                                        => Set.empty[String]
     }
 
-  def setupFastPassForUserInNewWorkspace(workspace: Workspace): ReadWriteAction[Unit] = {
+  def setupFastPassForUserInClonedWorkspace(parentWorkspace: Workspace, childWorkspace: Workspace): Future[Unit] = {
     if (!config.enabled) {
-      logger.debug(s"FastPass is disabled. Will not grant FastPass access to ${workspace.toWorkspaceName}")
-      return DBIO.successful()
+      logger.debug(s"FastPass is disabled. Will not grant FastPass access to ${parentWorkspace.toWorkspaceName}")
+      return Future.successful()
     }
-
-    try
-      DBIO.from(quotaAvailableForNewWorkspaceFastPassGrants(workspace)).flatMap { quotaAvailable =>
-        if (quotaAvailable) {
-          logger
-            .info(
-              s"Adding FastPass access for ${ctx.userInfo.userEmail.value} in workspace ${workspace.toWorkspaceName}"
-            )
-          val expirationDate = DateTime.now(DateTimeZone.UTC).plus(config.grantPeriod.toMillis)
-          for {
-            maybeUserStatus <- DBIO.from(samDAO.getUserStatus(ctx))
-            if maybeUserStatus.isDefined
-            samUserInfo = maybeUserStatus.map(SamUserInfo.fromSamUserStatus).orNull
-
-            roles <- DBIO
-              .from(samDAO.listUserRolesForResource(SamResourceTypeNames.workspace, workspace.workspaceId, ctx))
-            petEmail <- DBIO.from(samDAO.getUserPetServiceAccount(ctx, workspace.googleProjectId))
-            userType = getUserType(samUserInfo.userEmail)
-            userAndPet = UserAndPetEmails(samUserInfo.userEmail, userType, petEmail)
-            _ <- setupProjectRoles(workspace, roles, userAndPet, samUserInfo, expirationDate)
-            _ <- setupBucketRoles(workspace, roles, userAndPet, samUserInfo, expirationDate)
-            _ <- DBIO
-              .from(openTelemetry.incrementCounter("fastpass-granted-user", tags = openTelemetryTags).unsafeToFuture())
-          } yield ()
-        } else {
-          logger.info(
-            s"Not enough IAM Policy Role Binding quota available to add FastPass access for ${ctx.userInfo.userEmail.value} in workspace ${workspace.toWorkspaceName}"
-          )
-          DBIO.successful()
-        }
+    dataSource
+      .inTransaction { implicit dataAccess =>
+        for {
+          maybeUserStatus <- DBIO.from(samDAO.getUserStatus(ctx))
+          if maybeUserStatus.isDefined
+          samUserInfo = maybeUserStatus.map(SamUserInfo.fromSamUserStatus).orNull
+
+          petEmail <- DBIO.from(samDAO.getUserPetServiceAccount(ctx, childWorkspace.googleProjectId))
+          userType = getUserType(samUserInfo.userEmail)
+          userAndPet = UserAndPetEmails(samUserInfo.userEmail, userType, petEmail)
+          _ <- removeParentBucketReaderGrant(parentWorkspace, samUserInfo)
```

Comment:
Why do we need to do this for a cloned workspace?

Reply:
see: https://github.com/broadinstitute/rawls/pull/2326

Reply:
Basically, if a user clones a single workspace many times, they can run up against a quota that says there can only be 20 policy bindings for a user to the same role with different conditions. So, we just clear out the previous bucket reader grant and re-make it again with a new condition.